### PR TITLE
Cdd 1755: Add extra validation to the status endpoint. Also no validation of badgeID for exports

### PR DESCRIPTION
**Pull request by mattclark-zerogravit**

### CODE REVIEW

File: app/uk/gov/hmrc/customs/declaration/services/StatusResponseValidationService.scala [hunk 1]
```diff
@@ -27,29 +28,110 @@ import scala.xml.NodeSeq
...
+        declarationsLogger.debugWithoutRequestContext(s"Status response acceptanceDate failed validation $acceptanceDate")
+        Left(ErrorResponse.errorBadRequest(s"Declaration acceptance date is greater than ${declarationsConfigService.declarationsConfig.declarationStatusRequestDaysLimit} days old"))
+      } else{
+        Right(true)
+      }
+    })
   }
 
+  def extractTradeMovementType(declarationNode: NodeSeq): Either[ErrorResponse, String]  = {
+    val maybeExtractedTradeMovementType = extractField(declarationNode, "tradeMovementType").fold[Option[String]](None)(tradeMovementType => Some(tradeMovementType.head.substring(0,2)))
```
Comment (paulblakehmrc) [inline app/uk/gov/hmrc/customs/declaration/services/StatusResponseValidationService.scala]: What happens if tradeMovementType is empty or of length 1? Will it fail with an index out of bounds exception?

File: app/uk/gov/hmrc/customs/declaration/services/StatusResponseValidationService.scala [hunk 2]
```diff
@@ -27,29 +28,110 @@ import scala.xml.NodeSeq
...
 
-  def validate(xml: NodeSeq, badgeIdentifier: BadgeIdentifier): Boolean = {
+  val IMPORT_MOVEMENT_TYPE : String = "IM"
+  val EXPORT_MOVEMENT_TYPE : String = "EX"
+  val CO_MOVEMENT_TYPE : String = "CO"
+
+  val importProcedureCategories: Seq[String] = Seq( "40", "42", "61", "07", "51", "53", "71")
+  val exportProcedureCategories: Seq[String] = Seq( "10")
+
+ // tradeMovementType procedureCategory
```
Comment (paulblakehmrc) [inline app/uk/gov/hmrc/customs/declaration/services/StatusResponseValidationService.scala]: remove comment

File: app/uk/gov/hmrc/customs/declaration/services/StatusResponseValidationService.scala [hunk 3]
```diff
@@ -27,29 +28,110 @@ import scala.xml.NodeSeq
...
   }
 
-  private def extractBadgeIdentifier(communicationAddress: String) = {
-    communicationAddress.split(":").last
+  private def validateAcceptanceDate(declarationNode: NodeSeq): Either[ErrorResponse, Boolean] = {
+    extractField(declarationNode, "acceptanceDate").fold[Either[ErrorResponse, Boolean]]({
+      declarationsLogger.errorWithoutRequestContext("Status response acceptanceDate field is missing")
+      Left(ErrorResponse.errorBadRequest(s"Declaration acceptance date is greater than ${declarationsConfigService.declarationsConfig.declarationStatusRequestDaysLimit} days old"))
+    })(acceptanceDate => {
+      val parsedDateTime = ISODateTimeFormat.dateTimeNoMillis.withZoneUTC().parseDateTime(acceptanceDate.head)
```
Comment (paulblakehmrc) [inline app/uk/gov/hmrc/customs/declaration/services/StatusResponseValidationService.scala]: If the date is not the correct format it will fail with some parsing error. Should this be catered for?

File: app/uk/gov/hmrc/customs/declaration/services/StatusResponseValidationService.scala [hunk 4]
```diff
@@ -27,29 +28,110 @@ import scala.xml.NodeSeq
...
 
-  private def validateBadgeIdentifier(declarationNode: NodeSeq, badgeIdentifier: BadgeIdentifier): Boolean = {
-    extractField(declarationNode, "communicationAddress").fold(false)({ communicationAddress =>
-      badgeIdentifier.value == extractBadgeIdentifier(communicationAddress.head)
+  private def validateBadgeIdentifier(declarationNode: NodeSeq, badgeIdentifier: BadgeIdentifier): Either[ErrorResponse, Boolean] = {
+    extractField(declarationNode, "communicationAddress").fold[Either[ErrorResponse, Boolean]]({
+      declarationsLogger.errorWithoutRequestContext("Status response BadgeId field is missing")
+      Left(ErrorResponse.errorBadRequest("Badge Identifier is missing or invalid"))
+    })({ communicationAddress =>
+      if(badgeIdentifier.value != extractBadgeIdentifier(communicationAddress.head)){
```
Comment (paulblakehmrc) [inline app/uk/gov/hmrc/customs/declaration/services/StatusResponseValidationService.scala]: Should badgeIdentifier comparison be case-sensitive?

File: app/uk/gov/hmrc/customs/declaration/services/StatusResponseValidationService.scala [hunk 5]
```diff
@@ -27,29 +28,110 @@ import scala.xml.NodeSeq
...
+  private def isImportProcedureCategory(maybeProcedureCategory: Option[String]): Boolean = {
+    matchAgainstGivenCategories(maybeProcedureCategory, importProcedureCategories)
+  }
+
+  private def matchAgainstGivenCategories(maybeProcedureCategory: Option[String], categoriesToMatch: Seq[String]): Boolean ={
+    maybeProcedureCategory.exists(procedureCategory => categoriesToMatch.contains(procedureCategory))
+  }
+
+  private def extractProcedureCategory(declarationNode: NodeSeq): Option[String]  = {
+    extractField(declarationNode, "procedureCategory").fold[Option[String]](None)(procedureCategory => Some(procedureCategory.head.substring(0,2)))
```
Comment (paulblakehmrc) [inline app/uk/gov/hmrc/customs/declaration/services/StatusResponseValidationService.scala]: as above, how bomb-proof do we want to make the data extractions?

File: test/component/CustomsDeclarationSubmissionSpec.scala [hunk 6]
```diff
@@ -49,7 +49,7 @@ class CustomsDeclarationSubmissionSpec extends ComponentTestSpec with AuditServi
 
   private val apiSubscriptionKeyForXClientIdV3 = apiSubscriptionKeyForXClientIdV2.copy(version = VersionThree)
 
-  protected override val BadRequestErrorWith2Errors: String =
+  protected override val BadStatusResponseErrorInvalidDate: String =
```
Comment (paulblakehmrc) [inline test/component/CustomsDeclarationSubmissionSpec.scala]: Best to revert this back to its original name. Occurs in other specs too.

File: test/component/CustomsDeclarationStatusSpec.scala [hunk 7]
```diff
@@ -41,6 +43,7 @@ class CustomsDeclarationStatusSpec extends ComponentTestSpec with AuditService w
 
   private val mrn = "some-mrn"
   private val endpoint = s"/status-request/mrn/$mrn"
+//  private implicit val materializer: Materializer = app.materializer
```
Comment (paulblakehmrc) [inline test/component/CustomsDeclarationStatusSpec.scala]: remove comment (came from a merge I think)

File: test/component/CustomsDeclarationStatusSpec.scala [hunk 8]
```diff
@@ -49,16 +52,16 @@ class CustomsDeclarationStatusSpec extends ComponentTestSpec with AuditService w
 
   private val apiSubscriptionKeyForXClientIdV3 = apiSubscriptionKeyForXClientIdV2.copy(version = VersionThree)
 
-  private val validResponse =
-    """<stat:declarationManagementInformationResponse xmlns:stat="http://gov.uk/customs/declarations/status-request">
+  private def validResponse(acceptanceDateVal : String) =
+    raw"""<stat:declarationManagementInformationResponse xmlns:stat="http://gov.uk/customs/declarations/status-request">
```
Comment (paulblakehmrc) [inline test/component/CustomsDeclarationStatusSpec.scala]: `raw` typo

File: test/component/CustomsDeclarationStatusSpec.scala [hunk 9]
```diff
@@ -99,7 +102,7 @@ class CustomsDeclarationStatusSpec extends ComponentTestSpec with AuditService w
       status(result) shouldBe OK
 
       And("the response body is a valid status xml")
-      contentAsString(result) shouldBe validResponse
+      contentAsString(result) shouldBe validResponse(acceptanceDateVal.toString(DateTimeFormat.forPattern("yyyy-MM-dd'T'HH:mm:ss'Z'")))
```
Comment (paulblakehmrc) [inline test/component/CustomsDeclarationStatusSpec.scala]: can date pattern be replaced with `ISODateTimeFormat`

File: test/component/CustomsDeclarationStatusSpec.scala [hunk 10]
```diff
@@ -128,7 +131,7 @@ class CustomsDeclarationStatusSpec extends ComponentTestSpec with AuditService w
...
```
Comment (paulblakehmrc) [inline test/component/CustomsDeclarationStatusSpec.scala]: ditto

File: app/uk/gov/hmrc/customs/declaration/services/StatusResponseValidationService.scala (hunk 2 above)
Comment (mattclark-zerogravit) [reply app/uk/gov/hmrc/customs/declaration/services/StatusResponseValidationService.scala]: Will do

File: app/uk/gov/hmrc/customs/declaration/services/StatusResponseValidationService.scala (hunk 1 above)
Comment (mattclark-zerogravit) [reply app/uk/gov/hmrc/customs/declaration/services/StatusResponseValidationService.scala]: I'll create a helper function that will return None in these circumstances

File: app/uk/gov/hmrc/customs/declaration/services/StatusResponseValidationService.scala (hunk 5 above)
Comment (mattclark-zerogravit) [reply app/uk/gov/hmrc/customs/declaration/services/StatusResponseValidationService.scala]: Think you mean the empty or 1 character edge cases. I'll cover this off when I add add tests and function to protect against index out of bounds etc.

File: test/component/CustomsDeclarationStatusSpec.scala (hunk 7 above)
Comment (mattclark-zerogravit) [reply test/component/CustomsDeclarationStatusSpec.scala]: Will do

File: test/component/CustomsDeclarationStatusSpec.scala (hunk 10 above)
Comment (mattclark-zerogravit) [reply test/component/CustomsDeclarationStatusSpec.scala]: ok

File: test/component/CustomsDeclarationStatusSpec.scala (hunk 9 above)
Comment (mattclark-zerogravit) [reply test/component/CustomsDeclarationStatusSpec.scala]: ok

File: test/component/CustomsDeclarationStatusSpec.scala (hunk 8 above)
Comment (mattclark-zerogravit) [reply test/component/CustomsDeclarationStatusSpec.scala]: The raw is needed as otherwise the argument is not evaluated and ends up as just '$acceptanceDateVal'
